core/IssueQueue.hpp

```diff
@@ -0,0 +1,145 @@
+// <IssueQueue.hpp> -*- C++ -*-
+
+#pragma once
+
+#include "sparta/collection/Collectable.hpp"
+#include "sparta/events/EventSet.hpp"
+#include "sparta/events/StartupEvent.hpp"
+#include "sparta/events/UniqueEvent.hpp"
+#include "sparta/ports/DataPort.hpp"
+#include "sparta/ports/PortSet.hpp"
+#include "sparta/resources/Scoreboard.hpp"
+#include "sparta/simulation/Clock.hpp"
+#include "sparta/simulation/ParameterSet.hpp"
+#include "sparta/simulation/ResourceFactory.hpp"
+#include "sparta/simulation/TreeNode.hpp"
+#include "sparta/simulation/Unit.hpp"
+
+#include "sparta/resources/PriorityQueue.hpp"
+
+#include "CoreTypes.hpp"
+#include "ExecutePipe.hpp"
+#include "FlushManager.hpp"
+#include "Inst.hpp"
+
+namespace olympia
+{
+
+    /**
+     * \class IssueQueue
+     * \brief Defines the stages for an issue queue
+     */
+    class IssueQueue : public sparta::Unit
+    {
+        // Issue Queue interfaces between Dispatch and ExecutePipe and holds
+        // instructions until an available ExecutePipe becomes ready Dispatch -> Issue
+        // Queue -> ExecutePipe
+      public:
+        class IssueQueueParameterSet : public sparta::ParameterSet
+        {
+          public:
+            IssueQueueParameterSet(sparta::TreeNode* n) : sparta::ParameterSet(n) {}
+            PARAMETER(uint32_t, scheduler_size, 8, "Scheduler queue size")
+            PARAMETER(bool, in_order_issue, true, "Force in order issue")
+        };
+
+        class IssueQueueSorter
+        {
+          public:
+            IssueQueueSorter() {}
+
+            IssueQueueSorter(IssueQueueSorter* sorter) : dyn_sorter_(sorter) {}
+
+            bool operator()(const InstPtr existing, const InstPtr to_be_inserted) const
+            {
+                return dyn_sorter_->choose(existing, to_be_inserted);
+            }
+
+            bool choose(const InstPtr existing, const InstPtr to_be_inserted)
+            {
+                if (fifo_)
+                {
+                    // we just want to insert it in the back for fifo order
+                    return false;
+                }
+                else
+                {
+                    // age based insertion
+                    return existing->getUniqueID() > to_be_inserted->getUniqueID();
+                }
+            }
+
+            void setSorting(const bool sort_type) { fifo_ = sort_type; }
+
+          private:
+            bool fifo_ = true;
+            IssueQueueSorter* dyn_sorter_ = this;
+        };
+
+        IssueQueue(sparta::TreeNode* node, const IssueQueueParameterSet* p);
+        static const char name[];
+        void setExePipe(const std::string & exe_pipe_name, olympia::ExecutePipe* exe_pipe);
+        void setExePipeMapping(const InstArchInfo::TargetPipe tgt_pipe,
+                               olympia::ExecutePipe* exe_pipe);
+        typedef std::unordered_map<std::string, olympia::ExecutePipe*> StringToExePipe;
+
+        StringToExePipe const getExePipes() { return exe_pipes_; };
```

Comment:
`const` is in the wrong palce.  Also, have this return `const &` to avoid the copy

Reply:
Addressed in [d6bb2fa](https://github.com/riscv-software-src/riscv-perf-model/pull/140/commits/d6bb2fa8a225faa530aa74a712eb3eed0fb596dc)